oidc_provider/lib/utils/token.py

```diff
@@ -71,27 +69,25 @@ def encode_id_token(payload, client):
     Represent the ID Token as a JSON Web Token (JWT).
     Return a hash.
     """
-    keys = get_client_alg_keys(client)
-    _jws = JWS(payload, alg=client.jwt_alg)
-    return _jws.sign_compact(keys)
-
-
-def decode_id_token(token, client):
-    """
-    Represent the ID Token as a JSON Web Token (JWT).
-    Return a hash.
-    """
-    keys = get_client_alg_keys(client)
-    return JWS().verify_compact(token, keys=keys)
+    key = client.client_secret
+    if client.jwt_alg == 'RS256':
+        rsakeys = RSAKey.objects.all()
+        if not rsakeys:
+            raise Exception('You must have an RSA Key.')
+        rsakey = rsakeys[0]
+        key = serialization.load_pem_private_key(rsakey.key.encode(), None, default_backend()).private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.TraditionalOpenSSL,
+            encryption_algorithm=serialization.NoEncryption()).decode('utf8')
+    return jwt.encode(payload, key, algorithm=client.jwt_alg).decode()
 
 
 def client_id_from_id_token(id_token):
     """
     Extracts the client id from a JSON Web Token (JWT).
     Returns a string or None.
     """
-    payload = JWT().unpack(id_token).payload()
-    return payload.get('aud', None)
+    return jwt.decode(id_token, verify=False).get('aud', None)
```

Comment:
Basically the same thing: from an id_token extract the aud. We set verify to False, because the provider is not the audience of this JWT.